File: desktop/src/utils/alertInput.js
// Spec G2 §7/§9 — Eingabepruefung der Preis-Alarme (deutsches Komma, kein Tausenderpunkt).
// ZWILLING: android/app/src/main/java/com/example/yugiohscanner/ml/AlertInput.kt. Beide laufen gegen
// docs/fixtures/portfolio/alert-input.json. Wer eine Seite aendert, aendert beide.
const EURO = /^\d+([.,]\d{1,2})?$/;
const PCT = /^\d+([.,]\d)?$/;
const num = (t) => Number(t.replace(',', '.'));
const clean = (text) => String(text ?? '').trim();

// Zielpreis: leer heisst entfernen ({ value: null, error: null }).
export function parseTarget(text) {
  const t = clean(text);
  if (t === '') return { value: null, error: null };
  if (!EURO.test(t)) return { value: null, error: 'Ungültiger Betrag' };
  const v = num(t);
  return v > 0 ? { value: v, error: null } : { value: null, error: 'Betrag muss größer als 0 sein' };
}

export function parsePct(text) {
  const t = clean(text);
  if (!PCT.test(t)) return { value: null, error: 'Ungültige Zahl' };
  const v = num(t);
  return v >= 1 && v <= 500 ? { value: v, error: null } : { value: null, error: 'Prozent zwischen 1 und 500' };
}

export function parseMinEur(text) {
  const t = clean(text);
  if (!EURO.test(t)) return { value: null, error: 'Ungültiger Betrag' };
  return { value: num(t), error: null };
}

// Gespeicherter Betrag im Eingabefeld: 12.5 -> "12,5", 2 -> "2", null -> "".
export const toInput = (v) => (v == null ? '' : String(Number(v)).replace('.', ','));

File: android/app/src/main/java/com/example/yugiohscanner/ml/AlertInput.kt
package com.example.yugiohscanner.ml

/** Ergebnis einer Eingabe. value null und error null heisst "leer" (nur beim Zielpreis erlaubt: entfernen). */
data class AlertParse(val value: Double?, val error: String?)

/**
 * Spec G2 §7/§9 -- Eingabepruefung der Preis-Alarme (deutsches Komma, kein Tausenderpunkt).
 * ZWILLING: desktop/src/utils/alertInput.js. Beide laufen gegen docs/fixtures/portfolio/alert-input.json.
 * Wer eine Seite aendert, aendert beide.
 */
object AlertInput {
    private val EURO = Regex("""^\d+([.,]\d{1,2})?$""")
    private val PCT = Regex("""^\d+([.,]\d)?$""")

    private fun num(t: String): Double = t.replace(',', '.').toDouble()

    fun parseTarget(text: String?): AlertParse {
        val t = text.orEmpty().trim()
        if (t.isEmpty()) return AlertParse(null, null)
        if (!EURO.matches(t)) return AlertParse(null, "Ungültiger Betrag")
        val v = num(t)
        return if (v > 0) AlertParse(v, null) else AlertParse(null, "Betrag muss größer als 0 sein")
    }

    fun parsePct(text: String?): AlertParse {
        val t = text.orEmpty().trim()
        if (!PCT.matches(t)) return AlertParse(null, "Ungültige Zahl")
        val v = num(t)
        return if (v in 1.0..500.0) AlertParse(v, null) else AlertParse(null, "Prozent zwischen 1 und 500")
    }

    fun parseMinEur(text: String?): AlertParse {
        val t = text.orEmpty().trim()
        if (!EURO.matches(t)) return AlertParse(null, "Ungültiger Betrag")
        return AlertParse(num(t), null)
    }

    /** Wie JS String(Number(v)): 2.0 -> "2", 12.5 -> "12,5", null -> "". */
    fun toInput(v: Double?): String {
        if (v == null) return ""
        val s = if (v == Math.floor(v) && Math.abs(v) < 1e15) v.toLong().toString() else v.toString()
        return s.replace('.', ',')
    }
}
